cmd/tailscale/cli/ssh.go

```diff
@@ -138,25 +149,40 @@ func genKnownHosts(st *ipnstate.Status) []byte {
 		if len(ps.SSH_HostKeys) == 0 {
 			continue
 		}
-		// addEntries adds one line per each of p's host keys.
-		addEntries := func(host string) {
-			for _, hk := range ps.SSH_HostKeys {
-				hostKey := strings.TrimSpace(hk)
-				if strings.ContainsAny(hostKey, "\n\r") { // invalid
-					continue
-				}
-				fmt.Fprintf(&buf, "%s %s\n", host, hostKey)
+		host := strings.TrimSuffix(ps.DNSName, ".")
+
+		for _, hk := range ps.SSH_HostKeys {
+			hostKey := strings.TrimSpace(hk)
+			if strings.ContainsAny(hostKey, "\n\r") { // invalid
+				continue
 			}
+			fmt.Fprintf(&buf, "%s %s\n", host, hostKey)
 		}
-		if ps.DNSName != "" {
-			addEntries(ps.DNSName)
+	}
+	return buf.Bytes()
+}
+
+// nodeDNSNameFromArg returns the PeerStatus.DNSName value (but
+// without the trailing dot) from a peer in st that matches the input
+// arg (which can be a base name, full DNS name, or an IP)
+func nodeDNSNameFromArg(st *ipnstate.Status, arg string) (dnsName string, ok bool) {
+	if arg == "" {
+		return
+	}
+	argIP, _ := netaddr.ParseIP(arg)
+	for _, ps := range st.Peer {
+		dnsName = strings.TrimSuffix(ps.DNSName, ".")
+		if strings.EqualFold(strings.TrimSuffix(arg, "."), dnsName) {
+			return dnsName, true
 		}
-		if base, _, ok := strings.Cut(ps.DNSName, "."); ok {
-			addEntries(base)
+		if base, _, ok := strings.Cut(ps.DNSName, "."); ok && strings.EqualFold(base, arg) {
+			return dnsName, true
 		}
-		for _, ip := range st.TailscaleIPs {
-			addEntries(ip.String())
+		for _, ip := range ps.TailscaleIPs {
```

Comment:
reorder to bring this to the top? you can save a bunch of cuts and string comparisons. not that it matters much here :) 
and you can conditionally only look at IPs if `!argIP.IsZero()`